ruby/example_code/s3/auth_federation_token_request_test.rb

```diff
@@ -1,142 +1,88 @@
-# snippet-sourceauthor:[AWS]
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX - License - Identifier: Apache - 2.0
+
+# This code example allows a federated user with a limited set of
+# permissions to list the objects in the specified Amazon S3 bucket.
+
 # snippet-start:[s3.ruby.auth_federation_token_request_test.rb]
 require 'aws-sdk-s3'
 require 'aws-sdk-iam'
 
-USAGE = <<DOC
-
-Usage: ruby auth_federation_token_request_test.rb -b BUCKET -u USER [-r REGION] [-d] [-h]
-
-  Creates a federated policy for USER to list items in BUCKET for one hour.
-
-  BUCKET is required and must already exist.
-
-  USER is required and if not found, is created.
-
-  If REGION is not supplied, defaults to us-west-2.
-
-  -d gives you extra (debugging) information.
-
-  -h displays this message and quits.
-
-DOC
-
-def print_debug(debug, s)
-  if debug
-    puts s
-  end
-end
-
-# Get the user if they exist, otherwise create them
-def get_user(region, user_name, debug)
-  iam = Aws::IAM::Client.new(region: 'us-west-2')
-  
-  # See if user exists
-  user = iam.user(user_name)
- 
-  # If user does not exist, create them
-  if user == nil
-    user = iam.create_user(user_name: user_name)
-    iam.wait_until(:user_exists, user_name: user_name)
-    print_debug(debug, "Created new user #{user_name}")
-  else
-    print_debug(debug, "Found user #{user_name} in region #{region}")
+# Uses an AWS IAM user with a limited set of persmissions to list the objects
+# in the specified Amazon S3 bucket.
+# @param region [String] The AWS Region to use for configuring the AWS STS
+#   and S3 clients.
+# @param user_name [String] The user's name.
+# @param bucket_name [String] The bucket's name.
+# @return [Boolean] true if all operations succeed; otherwise, false.
+# @example
+#   list_objects_with_temporary_credentials('us-east-1', 'my-user', 'my-bucket')
+def list_objects_with_temporary_credentials(region, user_name, bucket_name)
+  # Create a new AWS STS client and get temporary credentials.
+  outcome = false
+  sts = Aws::STS::Client.new(region: region)
+  creds = ''
+
+  begin
+    creds = sts.get_federation_token({
+      duration_seconds: 3600,
+      name: user_name,
+      policy: "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"Stmt1\",\"Effect\":\"Allow\",\"Action\":\"s3:ListBucket\",\"Resource\":\"arn:aws:s3:::#{bucket_name}\"}]}"
+    })
+  rescue StandardError => e
+    puts "Error while getting federation token: #{e.message}"
   end
- 
- user
-end
-
-# main
-region = 'us-west-2'
-user_name = ''
-bucket_name = ''
-
-i = 0
-
-while i < ARGV.length
-  case ARGV[i]
-
-    when '-b'
-      i += 1
-      bucket_name = ARGV[i]
-
-    when '-u'
-      i += 1
-      user_name = ARGV[i]
-
-    when '-r'
-      i += 1
-      region = ARGV[i]
-
-    when '-h'
-      puts USAGE
-      exit 0
 
+  # Create an Amazon S3 resource with temporary credentials.
+  s3 = Aws::S3::Resource.new(region: region, credentials: creds)
```

Comment:
in general we're pushing away from Resource interfaces. please try to use the client API here (`Aws::S3::Client.new`)

resource interfaces are hand maintained and have limitations/bugs that can't be fixed because of their design. we're moving away from them.

For this code, you can do something like (also uses `max_keys` to limit the response):

```
s3 = Aws::S3::Client.new(region: region, credentials: creds)
resp = s3.list_objects_v2(bucket: bucket_name, max_keys: 50)
resp.contents.each do |obj|
  # obj.key, obj.etag, etc
end
```

https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/S3/Client.html#list_objects_v2-instance_method

Reply:
I changed this to Aws::S3::Client.new and used max_keys, as you propose.